pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/MailboxAssignmentVisitor.java

```diff
@@ -165,8 +165,13 @@ private static boolean isDirectExchangeCompatible(DispatchablePlanMetadata sende
       DispatchablePlanMetadata receiver) {
     int numSenders = sender.getWorkerIdToServerInstanceMap().size();
     int numReceivers = receiver.getWorkerIdToServerInstanceMap().size();
-    return numSenders * sender.getPartitionParallelism() == numReceivers && sender.getPartitionFunction() != null
-        && sender.getPartitionFunction().equalsIgnoreCase(receiver.getPartitionFunction());
+    if (numSenders * sender.getPartitionParallelism() != numReceivers) {
+      return false;
+    }
+    if (sender.getPartitionFunction() == null) {
```

Comment:
Without hint, when sender is pre-partitioned, does it always come with a partition function? If so, do we need to check receiver partition function?
Let's also add some comment on why we perform this check